src/dbt_osmosis/dbt_templater/templater.py

```diff
@@ -0,0 +1,228 @@
+"""Defines the dbt_osmosis templater."""
+
+import os.path
+import logging
+import threading
+from pathlib import Path
+from typing import Optional
+
+from dbt.version import get_installed_version
+from dbt.exceptions import (
+    CompilationException as DbtCompilationException,
+)
+from jinja2 import Environment
+from jinja2_simple_tags import StandaloneTag
+from sqlfluff.core.errors import SQLTemplaterError, SQLFluffSkipFile
+from sqlfluff.core.templaters.base import TemplatedFile, large_file_check
+from sqlfluff.core.templaters.jinja import JinjaTemplater
+
+from dbt_osmosis.core.osmosis import DbtProject
+
+templater_logger = logging.getLogger(__name__)
+
+DBT_VERSION = get_installed_version()
+DBT_VERSION_TUPLE = (int(DBT_VERSION.major), int(DBT_VERSION.minor))
+
+if DBT_VERSION_TUPLE >= (1, 3):
+    COMPILED_SQL_ATTRIBUTE = "compiled_code"
+    RAW_SQL_ATTRIBUTE = "raw_code"
+else:
+    COMPILED_SQL_ATTRIBUTE = "compiled_sql"
+    RAW_SQL_ATTRIBUTE = "raw_sql"
+
+local = threading.local()
```

Comment:
This is how SQLFluff works with external (plugin) templaters. Should I add this info to a "readme" file?

The templater exposes an entrypoint using `pyproject.toml`:
```
[tool.poetry.plugins."sqlfluff"]
"dbt_templater" = "dbt_osmosis.dbt_templater"
```

The entrypoint is implemented as a function `get_templaters()`. It returns a list of named templaters (typically just one). SQLFluff finds the templater whose `.name` field matches the one specified in `.sqlfluff`, then calls it to create the templater instance.

Because the templater is created by SQLFluff, if it needs to work with any `dbt-osmosis` state, it has to get that globally, and it's not inherently aware of `dbt-osmosis` threads or processes.

In order to do its job, the dbt templater needs access to the Jinja `Environment` _for the specific file that is being linted_. Hence, this may be a different object for each call to `/lint`. For simplicity and safety when running in a server environment, I made a couple of decisions:
* Monkeypatch `Environment.from_string()` globally at startup. This avoids the risk and potential race conditions of multiple threads potentially installing and removing patches concurrently.
* However, because (as noted above), each `/lint` call potentially uses a different `Environment` **instance**, I used thread-local storage to keep those separate.

Details on how thread-local storage is used:
1. `_unsafe_process()` is called during the processing of a specific SQL file. It sets `local.make_template = None` and `local.target_sql` to the SQL path or string being processed in the current thread.
2. Next, `_unsafe_process()` calls `osmosis_dbt_project.compile_sql.__wrapped_()` to get dbt to compile the SQL. It bypasses the LRU cache in order to force compilation.
3. When dbt compiles the SQL, it internally calls the patched version of `Environment.from_string()`. The patched version checks the function parameters to determine if it's compiling the SQL being linted, or some other SQL. If it's the SQL being linted, it creates a `make_template()` function that uses the Jinja `Environment` and "exports" that function by setting `local.make_template`.
4. After dbt finishes compiling the SQL, the `_unsafe_process()` function calls `self.slice_file()`, passing `local.make_template` as one of the parameters. `self.slice_file()` uses this function to create and compile a second Jinja template, similar to the SQL being linted, but with special markers that allow SQLFluff to map between the uncompiled and compiled SQL. This is how SQLFluff manages to lint and fix templated SQL in "uncompiled" space.